### PR TITLE
os/KeyValueStore.cc : remove keyvaluestore queue and thread

### PR DESCRIPTION
**Pull request by xinxinsh**
make write path short for keyvaluestore by removing KeyValueStore queue and thread

Signed-off-by: xinxin shu <xinxin.shu@intel.com>

### CODE REVIEW

File: src/os/KeyValueStore.cc [hunk 1]
```diff
@@ -530,10 +530,6 @@ KeyValueStore::KeyValueStore(const std::string &base,
   op_queue_len(0), op_queue_bytes(0),
   op_throttle_lock("KeyValueStore::op_throttle_lock"),
   op_finisher(g_ceph_context),
-  op_tp(g_ceph_context, "KeyValueStore::op_tp",
```
Comment (prallabh) [inline src/os/KeyValueStore.cc]: Does this call for removal of configurable values keyvaluestore_op_threads and keyvaluestore_op_thread_suicide_timeout from config_opts.h?

**Comment by yuyuyu101** [comment]
I think we need to have a separated thread to do transaction submit if we don't use thread pool for doing transactions. It will result in less throughput because pg thread will block in submitting transaction.

**Comment by xinxinsh** [comment]
@yuyuyu101 ,  i test this on my 2-nodes SSD setup(16 osds cluster), with this pr, we can get ~3x performance for 4K random write

**Comment by xinxinsh** [comment]
@prallabh ， updated

File: src/os/KeyValueStore.cc [hunk 2]
```diff
@@ -1026,6 +1020,8 @@ int KeyValueStore::queue_transactions(Sequencer *posr, list<Transaction*> &tls,
     dump_transactions(o->tls, o->op, osr);
   dout(5) << "queue_transactions (trailing journal) " << " " << tls <<dendl;
   queue_op(osr, o);
+  do_op(osr, *handle);
+  finish_op(osr);
```
Comment (varadakari) [inline src/os/KeyValueStore.cc]: I have a similar patch, which eliminates the osr also. it completes the request in same thread context(sharded worker thread),

**Comment by xinxinsh** [comment]
@varadakari ,  if you remove osr, how do you prevent concurrency write on one pg?

**Comment by varadakari** [comment]
we will have the pg lock held right? Anyways write completion in a  finisher thread.

**Comment by yuyuyu101** [comment]
Hmm, if we do this online, we could discard osr totally

**Comment by yuyuyu101** [comment]
@varadakari do you implement a dedicated thread to submit transaction? What's the effects?

**Comment by xinxinsh** [comment]
@varadakari , yes, thanks for pointing me out

**Comment by varadakari** [comment]
@yuyuyu101 Nope, increased the number of threads and shards. So far haven't seen any problems. Calling the completions, queuing to finisher etc.. are done in the same thread.  I have a patch for these changes, will send a PR soon.

**Comment by xinxinsh** [comment]
@varadakari , do you mean pg lock at https://github.com/ceph/ceph/blob/master/src/osd/OSD.cc#L8198 , if yes, i think there are other ops that directly invoke store->queue_transaction, not go through OSD::ShardedOpWQ::_process, how do you prevent these ops and client ops operating on the same pg?

**Comment by varadakari** [comment]
@xinxinsh Sorry didn't get any notification for you comment. Yes i was referring to the same lock. I am not sure if any op can do a queue_transaction() by not coming through ShardedOpWQ. If this is a primary osd, we will have a object level lock till we commit the changes to serve the request. If that is a replica op, that will come through ShardedOpWQ. Not sure if i am missing any more details.

**Comment by dachary** [comment]
@xinxinsh this needs rebasing

**Pull request closed without merging**